src/rendezvous/channels/MSC3903ECDHv1RendezvousChannel.ts

```diff
@@ -0,0 +1,279 @@
+ */
+export class MSC3903ECDHv1RendezvousChannel implements RendezvousChannel {
+    private olmSAS?: SAS;
+    private ourPublicKey: Uint8Array;
+    private aesKey?: CryptoKey | Uint8Array;
+
+    public constructor(
+        public transport: RendezvousTransport,
+        private theirPublicKey?: Uint8Array,
+        public onFailure?: (reason: RendezvousFailureReason) => void,
+    ) {
+        this.olmSAS = new global.Olm.SAS();
+        this.ourPublicKey = decodeBase64(this.olmSAS.get_pubkey());
+    }
+
+    public async generateCode(intent: RendezvousIntent): Promise<ECDHv1RendezvousCode> {
+        if (this.transport.ready) {
+            throw new Error('Code already generated');
+        }
+
+        await this.send({ algorithm: RendezvousChannelAlgorithm.ECDH_V1 });
+
+        const rendezvous: ECDHv1RendezvousCode = {
+            "rendezvous": {
+                algorithm: RendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+                transport: await this.transport.details(),
+            },
+            intent,
+        };
+
+        return rendezvous;
+    }
+
+    public async connect(): Promise<string> {
+        if (!this.olmSAS) {
+            throw new Error('Channel closed');
+        }
+
+        const isInitiator = !this.theirPublicKey;
+
+        if (isInitiator) {
+            // wait for the other side to send us their public key
+            const res = await this.receive();
+            if (!res) {
+                throw new Error('No response from other device');
+            }
+            const { key, algorithm } = res;
+
+            if (algorithm !== RendezvousChannelAlgorithm.ECDH_V1 || !key) {
+                throw new RendezvousError(
+                    'Unsupported algorithm: ' + algorithm,
+                    RendezvousFailureReason.UnsupportedAlgorithm,
+                );
+            }
+
+            this.theirPublicKey = decodeBase64(key);
+        } else {
+            // send our public key unencrypted
+            await this.send({
+                algorithm: RendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+            });
+        }
+
+        this.olmSAS.set_their_key(encodeBase64(this.theirPublicKey!));
+
+        const initiatorKey = isInitiator ? this.ourPublicKey : this.theirPublicKey!;
+        const recipientKey = isInitiator ? this.theirPublicKey! : this.ourPublicKey;
+        let aesInfo = RendezvousChannelAlgorithm.ECDH_V1.toString();
+        aesInfo += `|${encodeBase64(initiatorKey)}`;
+        aesInfo += `|${encodeBase64(recipientKey)}`;
+
+        const aesKeyBytes = this.olmSAS.generate_bytes(aesInfo, 32);
+
+        this.aesKey = await importKey(aesKeyBytes);
+
+        const rawChecksum = this.olmSAS.generate_bytes(aesInfo, 5);
+        return generateDecimalSas(Array.from(rawChecksum)).join('-');
+    }
+
+    private async encrypt(data: any): Promise<string> {
```

Comment:
We'll probably eventually want to factor this function (and `decrypt`) out into `crypto/aes.ts`, but it should be OK to keep it here for now.

Reply:
Agreed. You had mentioned that you wanted to migrate to AES-GCM for the file attachments at some point. I figured that would be a good time to do it.